### PR TITLE
Background Worker template

### PR DESCRIPTION
**Pull request by Tratcher**
This is a first pass at #6817. @glennc I copied in a lot of your prototype, perhaps more than we're ready for this time around. The basic empty version works. Do we need to sort out the helm, webjobs, and docker stuff for preview3?

### CODE REVIEW

**Comment by Tratcher** [comment]
WebJobs does seem to be working locally.

File: src/ProjectTemplates/Web.ProjectTemplates/Worker-CSharp.csproj.in [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  <PropertyGroup>
+    <TargetFramework>netcoreapp3.0</TargetFramework>
+    <OutputType>Exe</OutputType>
+    <NoDefaultLaunchSettingsFile Condition="'$(ExcludeLaunchSettings)' == 'True'">True</NoDefaultLaunchSettingsFile>
+    <RootNamespace Condition="'$(name)' != '$(name{-VALUE-FORMS-}safe_namespace)'">Company.Application1</RootNamespace>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <!--#if (WebJobs) -->
+    <PackageReference Include="Microsoft.Azure.WebJobs" Version="3.0.2" /> <!--TODO Version-->
```
Comment (natemcmaster) [inline src/ProjectTemplates/Web.ProjectTemplates/Worker-CSharp.csproj.in]: Should this be managed via `eng/Versions.props`?

Comment (Tratcher) [reply src/ProjectTemplates/Web.ProjectTemplates/Worker-CSharp.csproj.in]: I expect so. I wanted to make sure it worked first.

**Comment by natemcmaster** [comment]
Build infrastructure looks fine. I don't have any other feedback on this. Don't have enough context on this PR to sign off or request changes.

cc @ryanbrandenburg, our project templating expert

**Comment by rynowak** [comment]
@stevejgordon

**Comment by Tratcher** [comment]
Hmm, VS doesn't pick up locally built and installed versions of the templates. Is that supposed to work?

**Comment by rynowak** [comment]
> Is that supposed to work?

I don't think so. I think it only honours the SDK.

File: src/ProjectTemplates/Web.ProjectTemplates/Worker-CSharp.csproj.in [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    <TargetFramework>netcoreapp3.0</TargetFramework>
+    <OutputType>Exe</OutputType>
+    <NoDefaultLaunchSettingsFile Condition="'$(ExcludeLaunchSettings)' == 'True'">True</NoDefaultLaunchSettingsFile>
+    <RootNamespace Condition="'$(name)' != '$(name{-VALUE-FORMS-}safe_namespace)'">Company.Application1</RootNamespace>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <!--#if (WebJobs) -->
+    <PackageReference Include="Microsoft.Azure.WebJobs" Version="${MicrosoftAzureWebJobsPackageVersion}" />
+    <PackageReference Include="Microsoft.Azure.WebJobs.Extensions" Version="${MicrosoftAzureWebJobsExtensionsPackageVersion}" />
```
Comment (davidfowl) [inline src/ProjectTemplates/Web.ProjectTemplates/Worker-CSharp.csproj.in]: Get rid of the webjobs support, we’re not shipping it

**Comment by davidfowl** [comment]
Remove webjobs, helm and docker. Let get the skeleton in then decide how we do the cross cutting concerns.

File: src/ProjectTemplates/Web.ProjectTemplates/content/Worker-CSharp/Worker.cs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+        public Worker(ILogger<Worker> logger)
+        {
+            _logger = logger;
+        }
+
+#if Empty
+        protected override async Task ExecuteAsync(CancellationToken stoppingToken)
+        {
+            while(!stoppingToken.IsCancellationRequested)
```
Comment (rynowak) [inline src/ProjectTemplates/Web.ProjectTemplates/content/Worker-CSharp/Worker.cs]: space

**Review by glennc — CHANGES_REQUESTED** [review]
Remove all the options for now, we'll go through and add them back 1 by 1 thinking about each one as we do. Need to consider partners for all of them, mostly all the VS experiences around HELM and Docker still working.

File: src/ProjectTemplates/Web.ProjectTemplates/content/Worker-CSharp/Worker.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+        {
+            _logger = logger;
+        }
+
+        protected override async Task ExecuteAsync(CancellationToken stoppingToken)
+        {
+            while (!stoppingToken.IsCancellationRequested)
+            {
+                _logger.LogInformation($"Worker running at: {DateTime.Now}");
+                await Task.Delay(1000);
```
Comment (davidfowl) [inline src/ProjectTemplates/Web.ProjectTemplates/content/Worker-CSharp/Worker.cs]: pass the token into the delay.

File: src/ProjectTemplates/Web.ProjectTemplates/content/Worker-CSharp/Program.cs [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+    public class Program
+    {
+        public static void Main(string[] args)
+        {
+            CreateHostBuilder(args).Build().Run();
+        }
+
+        public static IHostBuilder CreateHostBuilder(string[] args) =>
+            Host.CreateDefaultBuilder(args)
+                .ConfigureServices(services => {
```
Comment (davidfowl) [inline src/ProjectTemplates/Web.ProjectTemplates/content/Worker-CSharp/Program.cs]: Fix the formatting to match the other templates.

File: src/ProjectTemplates/Web.ProjectTemplates/content/Worker-CSharp/.template.config/template.json [hunk 6]
```diff
@@ -0,0 +1,90 @@
+{
+  "$schema": "http://json.schemastore.org/template",
+  "author": "Microsoft",
+  "classifications": [
+    "Common",
+    "Worker"
+  ],
+  "name": ".NET Worker",
+  "generatorVersions": "[1.0.0.0-*)",
+  "description": "An empty project template for creating an .NET worker application. This template does not have any content in it.",
```
Comment (davidfowl) [inline src/ProjectTemplates/Web.ProjectTemplates/content/Worker-CSharp/.template.config/template.json]: > An empty project template for creating an .NET worker application. This template does not have any content in it.

This should be _a .NET worker_ not _an_

**Comment by Tratcher** [comment]
@ryanbrandenburg @rynowak @glennc if `dotnet new -i` doesn't work for VS then how to we verify `vs-2017.3.host.json`?

**Comment by Tratcher** [comment]
I was able to get VS to detect locally built templates by pasting Microsoft.DotNet.Web.ProjectTemplates.3.0.3.0.0-preview-t000.nupkg into `D:\github\AspNetCore\.dotnet\x64\sdk\3.0.100-preview-009750\Templates`  and deleting the prior version. @glennc new the worker template does not show up in the oneaspnet dialog when I do this. Is it filtered by classification?

File: src/ProjectTemplates/Web.ProjectTemplates/content/Worker-CSharp/.template.config/template.json [hunk 7]
```diff
@@ -0,0 +1,90 @@
+{
+  "$schema": "http://json.schemastore.org/template",
+  "author": "Microsoft",
+  "classifications": [
+    "Common",
+    "Worker"
+  ],
+  "name": "Worker Service",
+  "generatorVersions": "[1.0.0.0-*)",
+  "description": "An empty project template for creating a worker service. This template does not have any content in it.",
```
Comment (glennc) [inline src/ProjectTemplates/Web.ProjectTemplates/content/Worker-CSharp/.template.config/template.json]: I don't think we need to say that 'This template does not have any content in it.' Unless you had a reason for it?

File: src/ProjectTemplates/Web.ProjectTemplates/content/Worker-CSharp/.template.config/vs-2017.3.host.json [hunk 8]
```diff
@@ -0,0 +1,39 @@
+{
+  "$schema": "http://json.schemastore.org/vs-2017.3.host",
+  "name": {
+    "text": "Worker Service",
+    "package": "{0CD94836-1526-4E85-87D3-FB5274C5AFC9}",
+    "id": "1011"
+  },
+  "description": {
+    "text": "An empty project template for creating a worker service. This template does not have any content in it.",
```
Comment (glennc) [inline src/ProjectTemplates/Web.ProjectTemplates/content/Worker-CSharp/.template.config/vs-2017.3.host.json]: As above.

File: src/ProjectTemplates/Web.ProjectTemplates/content/Worker-CSharp/.template.config/vs-2017.3.host.json [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    "package": "{0CD94836-1526-4E85-87D3-FB5274C5AFC9}",
+    "id": "1012"
+  },
+  "order": 100,
+  "icon": "vs-2017.3/Empty.png",
+  "learnMoreLink": "https://go.microsoft.com/fwlink/?LinkID=784883",
```
Comment (glennc) [inline src/ProjectTemplates/Web.ProjectTemplates/content/Worker-CSharp/.template.config/vs-2017.3.host.json]: use this: 'https://go.microsoft.com/fwlink/?linkid=2072778' we will update it to point to what we want later.

File: src/ProjectTemplates/Web.ProjectTemplates/content/Worker-CSharp/.template.config/vs-2017.3.host.json [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+  "uiFilters": [ "oneaspnet" ],
+  "supportsDocker": true,
+  "legacyTemplateIdentity": "Microsoft.NetCore.CSharp.EmptyWorker",
```
Comment (glennc) [inline src/ProjectTemplates/Web.ProjectTemplates/content/Worker-CSharp/.template.config/vs-2017.3.host.json]: All this auth and port stuff is presumably not required.

**Review by glennc — DISMISSED** [review]

File: src/ProjectTemplates/Web.ProjectTemplates/content/Worker-CSharp/.template.config/template.json (hunk 7 above)
Comment (Tratcher) [reply src/ProjectTemplates/Web.ProjectTemplates/content/Worker-CSharp/.template.config/template.json]: It was copied from Empty. This is the worker equivalent of Empty. Keep or kill?

File: src/ProjectTemplates/Web.ProjectTemplates/content/Worker-CSharp/.template.config/vs-2017.3.host.json (hunk 10 above)
Comment (Tratcher) [reply src/ProjectTemplates/Web.ProjectTemplates/content/Worker-CSharp/.template.config/vs-2017.3.host.json]: I'm not messing with the VS template until we can test it in VS. This is a copy of Empty.

**Comment by Tratcher** [comment]
Updated. Anybody else?

**Review by davidfowl — APPROVED** [review]
LGTM. Now we need to try it out once it goes in. I hope we don't generate IIS launch settings for this...

**Comment by davidfowl** [comment]
SHIPIT!

**Pull request merged**